### DISCUSSION TITLE
Detailing requirements for new and improved MCTS

### DISCUSSION THREAD

**Discussion by oscardssmith** [LeelaChessZero/lc0#1591] [Ideas]
- General Goals

1. Policy should only influence behavior for nodes with few visits
2. Monotonic
3. Fewer hyper-parameters
4. FPU sort of feels like a hack. The only actual information that we have at 0 visits is the P, so the term that deals with policy shouldn't need a special case at n=0

- At Root

1. Simple regret minimization

Current best effort
At root

![f1]

In tree

![f2]

[f1]: https://cdn.discordapp.com/attachments/689595983419670571/702694343051313172/635246878912413748.png

[f2]: https://cdn.discordapp.com/attachments/689595983419670571/702694572664422571/635246878912413748.png

**Comment by AlexisOlson** [comment]
Here's where the U term is currently implemented:

https://github.com/LeelaChessZero/lc0/blob/1eb0d56e44b11082e4de634e78fb22328108aeef/src/mcts/node.h#L397-L401

Here's where the passed argument is computed:

https://github.com/LeelaChessZero/lc0/blob/1eb0d56e44b11082e4de634e78fb22328108aeef/src/mcts/search.cc#L1028-L1029

**Comment by farmersrice** [comment]
> Fewer hyper-parameters

This should probably be relaxed to fewer hyperparameters OR equal hyperparameters and decent strength gain.

Also it might be the case that pst can be removed (set to 1) with the new formulas, since changing pst = changing the sharpness of policies. In the new formula, the behavior you could get from pst is approximately the behavior you would get from changing c_1 and c_3 (policies becoming more or less important in the search).

**Comment by kiudee** [comment]
The formulas in the first post are supposed to minimize simple and cumulative regret. Can we list the corresponding scientific papers?

**Comment by oscardssmith** [comment]
Good idea. Will do.

**Comment by Naphthalin** [comment]
[https://link.springer.com/content/pdf/10.1007/s10472-011-9258-6.pdf](https://link.springer.com/content/pdf/10.1007/s10472-011-9258-6.pdf) is the paper given as the source of the current PUCT formula in the original AlphaGo paper.

**Comment by kiudee** [comment]
The original PUCT formula slightly resembles the one posted above, with a few deviations:
![Screenshot_2020-04-23 s10472-011-9258-6 pdf](https://user-images.githubusercontent.com/564473/80120432-46067c80-858b-11ea-8596-a567622ae0e9.png)
The variable Mi here is the policy of move i.

**Comment by isty2e** [comment]
Here is a slightly modified version of the formula, whose purpose is to have a few tunable parameters and removal of if-statements, and monotonicity of the m term.
![puct](https://user-images.githubusercontent.com/571680/80123632-1fb40180-85ca-11ea-99d4-9526783610d0.png)
c_att is added for numerical stability for p~0 cases, but can be removed. Probably this formula can be enhanced further for low-visit cases.

**Comment by Naphthalin** [comment]
Other PRs and issues concerning the PUCT formula and policies:

#796 #818 #791 #913 #918 #746 #743 #699 #1235 #1173 #1150

**Comment by Naphthalin** [comment]
As you can see I tried to collect all issues and PRs so far which discuss similar things, some of them showing that the people who have posted here so far aren't happy with the PUCT formula for quite a while :)

@oscardssmith what do you suggest on how to proceed here?
